woocommerce/woocommerce-ios
# Image Upload: make an image request cancellable

"Allow an image request to be cancelled when used in a reusable cell" for #1713 

## Changes

- Before this PR, `ImageDownloader` is a class of `Kingfisher` that downloads an image. This PR created an `ImageDownloader` protocol and made `Kingfisher.ImageDownloader` conform to the protocol
- In `ImageService` protocol, returned a `ImageDownloadTask` for an image download request
  - `ImageDownloadTask` conforms to a `Cancellable` protocol (built-in iOS 13+ but we can't use it yet)
- In `ProductUIImageLoader` protocol, returned a `Cancellable` for a UI image request similar to `ImageService`
  - In its implementation `DefaultProductUIImageLoader`, kept track of the active image requests and cancelled them on `deinit`
- In `ProductImageCollectionViewCell`, added a property of `Cancellable` and cancelled the optional task on `prepareForReuse`
  - Updated its usage in `ProductImagesCollectionViewController` and `ProductImagesCollectionViewDataSource`
- In `AztecTextViewAttachmentHandler`, kept track of the active image requests and cancelled them on `deinit`
- In WP media library data source (`WordPressMediaLibraryPickerDataSource`), created a new class `CancellableMedia` that includes a `Media` and `Cancellable` so that a media can be cancelled to replace the media items used in `WordPressMediaLibraryPickerDataSource`
  - Reverted `Media` back to a struct!
- Updated unit tests

## Testing

It's a bit harder to test this change, but it's possible in WP media library picker with `Wormholy`:

- Please make sure the image cache for the WP media library is empty, like reinstalling the app or switching to a different store that you haven't opened the WP media library before.
- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- (Please check the bold text before tapping the option) Tap "WordPress Media Library", then **make sure you've seen some cells whose image haven't been loaded and quickly leave the screen**
- Leave edit product screen, and go to "My store" tab
- Tap settings > "Launch Wormholy debug" and filter by "uploads" for image requests --> there should be some requests that never complete (cancelled) like in the screenshot below (you can try the same steps where you leave the picker before some cell images are loaded on `develop`, and all image requests complete)
 
`develop` | PR branch
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-04-09 at 14 17 19](https://user-images.githubusercontent.com/1945542/78866597-47578580-7a72-11ea-96e9-f2455836c2d2.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-09 at 14 22 25](https://user-images.githubusercontent.com/1945542/78866437-f5aefb00-7a71-11ea-8ef1-ece7853749e8.png)


Update release notes:
